src/optimizer/projection_push_down_optimizer.cpp

```diff
@@ -0,0 +1,134 @@
+#include "optimizer/projection_push_down_optimizer.h"
+
+#include "planner/logical_plan/logical_operator/logical_accumulate.h"
+#include "planner/logical_plan/logical_operator/logical_filter.h"
+#include "planner/logical_plan/logical_operator/logical_hash_join.h"
+#include "planner/logical_plan/logical_operator/logical_intersect.h"
+#include "planner/logical_plan/logical_operator/logical_order_by.h"
+#include "planner/logical_plan/logical_operator/logical_projection.h"
+
+using namespace kuzu::common;
+using namespace kuzu::planner;
+using namespace kuzu::binder;
+
+namespace kuzu {
+namespace optimizer {
+
+void ProjectionPushDownOptimizer::rewrite(planner::LogicalPlan* plan) {
+    visitOperator(plan->getLastOperator().get());
+}
+
+void ProjectionPushDownOptimizer::visitOperator(LogicalOperator* op) {
+    switch (op->getOperatorType()) {
+    case LogicalOperatorType::ACCUMULATE: {
+        visitAccumulate(op);
+    } break;
+    case LogicalOperatorType::FILTER: {
+        visitFilter(op);
+    } break;
+    case LogicalOperatorType::HASH_JOIN: {
+        visitHashJoin(op);
+    } break;
+    case LogicalOperatorType::PROJECTION: {
+        visitProjection(op);
+        return;
+    }
+    case LogicalOperatorType::INTERSECT: {
+        visitIntersect(op);
+    } break;
+    case LogicalOperatorType::ORDER_BY: {
+        visitOrderBy(op);
+    } break;
+    default:
+        break;
+    }
+    for (auto i = 0; i < op->getNumChildren(); ++i) {
+        visitOperator(op->getChild(i).get());
+    }
+}
+
+void ProjectionPushDownOptimizer::visitAccumulate(planner::LogicalOperator* op) {
+    auto accumulate = (LogicalAccumulate*)op;
+    for (auto& expression : accumulate->getExpressions()) {
+        collectPropertiesInUse(expression);
+    }
+}
+
+void ProjectionPushDownOptimizer::visitFilter(planner::LogicalOperator* op) {
+    auto filter = (LogicalFilter*)op;
+    collectPropertiesInUse(filter->getPredicate());
+}
+
+void ProjectionPushDownOptimizer::visitHashJoin(planner::LogicalOperator* op) {
+    auto hashJoin = (LogicalHashJoin*)op;
+    for (auto& joinNodeID : hashJoin->getJoinNodeIDs()) {
+        collectPropertiesInUse(joinNodeID);
+    }
+    if (hashJoin->getJoinType() == JoinType::MARK) { // no need to perform push down for mark join.
+        return;
+    }
+    expression_set expressionsBeforePruning;
```

Comment:
It seems you don't need the `expressionsBeforePruning`.
You can loop over hashJoin->getExpressionsToMaterialize() and directly insert into `expressionsAfterPruning` when necessary.